### ISSUE TITLE
Codec.from changes decoder behaviour

### ISSUE THREAD

**Issue by jakubdaniel** [circe/circe#2104]
Assuming
```scala
val e = io.circe.Encoder.instance[Option[Int]] { _ => ??? } // fake, not used in this example
val d = io.circe.Decoder.decodeOption[Int]
val c = io.circe.Codec.from(d, e)
val h = io.circe.Json.obj().hcursor
```
is
```scala
scala> d.at("x").tryDecode(h)
val res0: io.circe.Decoder.Result[Option[Int]] = Right(None)

scala> c.at("x").tryDecode(h)
val res1: io.circe.Decoder.Result[Option[Int]] = Left(DecodingFailure at .x: Missing required field)
```
expected?

I'd expect `c` to behave exactly as `d` on all inputs to any of the `Decoder` methods. But it obviously doesn't. This is because it falls back to the default implementations of `tryDecode` etc instead of properly delegating all the methods to any potential overrides already baked into `d`.

**Comment by zarthross** [comment]
Looks to me like the `Codec.from` method only overrides the 2 abstract methods from Encoder and Decoder, but in reality, we override various other methods from Encoder/Decoder all the time... we need to update that `Codec.from` implementation (and any other implementations of `Codec` for that matter) so that it correctly proxies ALL methods to the correct underlying `Encoder/Decoder`

**Comment by satorg** [comment]
It seems, this one was resolved via #2131 , wasn't it?
